sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Misc/MiscellaneousTests.cs

```diff
@@ -0,0 +1,245 @@
+                .Concat(NuGetMetadataReference.MicrosoftEntityFrameworkCoreSqlServer(entityFrameworkVersion))
+                .Concat(NuGetMetadataReference.MicrosoftEntityFrameworkCoreRelational(entityFrameworkVersion));
+
+        private Type[] GetSonarAnalyzerTypes(Assembly assembly)
+        {
+            var allAnalyzerTypes = assembly.GetTypes()
+                .Where(IsSonarAnalyzer).ToArray();
+
+            Log($"Analyzer count: {allAnalyzerTypes.Length}");
+
+            return allAnalyzerTypes;
+        }
+
+        private bool IsSonarAnalyzer(Type type) =>
+            (!type.IsAbstract && IsDerivedFromSonarDiagnosticAnalyzer(type));
+
+        private static readonly Type SonarDiagnosticAnalyzerType = typeof(SonarDiagnosticAnalyzer);
+
+        private bool IsDerivedFromSonarDiagnosticAnalyzer(Type type)
+        {
+            var current = type.BaseType;
+
+            while (current != null)
+            {
+                if (current == SonarDiagnosticAnalyzerType)
+                {
+                    return true;
+                }
+                current = current.BaseType;
+            }
+            return false;
+        }
+
+        private SonarDiagnosticAnalyzer CreateAnalyzer(Type analyzerType) =>
+            Activator.CreateInstance(analyzerType) as SonarDiagnosticAnalyzer;
+
+        private bool ExecuteAnalyzerWithTimeout(Compilation compilation, SonarDiagnosticAnalyzer analyzer, int timeoutMs)
+        {
+            var cancellationSource = new CancellationTokenSource(timeoutMs);
+
+            try
+            {
+                DiagnosticVerifier.GetAllDiagnostics(compilation, new[] { analyzer }, CompilationErrorBehavior.FailTest, cancellationSource.Token);
+            }
+            catch(AggregateException ex)
+            {
+                if (ex.InnerExceptions.Any(x => x is OperationCanceledException))
+                {
+                    return false;
+                }
+                throw ex;
+            }
+            return true;
+        }
+
+        private void AssertNoUnexpectedSlowAnalyzers(IEnumerable<ExecutionResult> executionResults, IEnumerable<string> expectedSlowAnalyzers)
+        {
+            var actualSlowAnalyzers = executionResults.Where(x => x.TimedOut).Select(x => x.AnalyzerType.FullName);
+            AssertNoUnexpectedSlowAnalyzers(actualSlowAnalyzers.Except(expectedSlowAnalyzers, StringComparer.OrdinalIgnoreCase));
+
+            var unexpectedlyFastAnalyzers = expectedSlowAnalyzers.Except(actualSlowAnalyzers, StringComparer.OrdinalIgnoreCase);
+
+            if (!unexpectedlyFastAnalyzers.Any())
+            {
+                Log(FormatListToMessage(expectedSlowAnalyzers, "The expected analyzers were slow:"));
+                return;
+            }
+
+            string unexpectedlyFastMessage = FormatListToMessage(unexpectedlyFastAnalyzers,
+                "The following analyzers where unexpectedly fast - the baseline may need to be updated:");
```

Comment:
would it be better : `The following analyzers where expected to be slow, but performed better than expected - the baseline may need to be updated:`